### ISSUE TITLE
JAWS React Front-End Module: Design Specs & Help Wanted

### ISSUE THREAD

**Issue by ac360** [serverless/serverless#41]
**This thread is for discussing the `JAWS-site-react` module's design and finding developers who want to help build it**

Overall, JAWS should support all front-end frameworks.  There has been a lot of demand for JAWS + React support.  So, let's start with React!

JAWS already comes with a `site` module.  But you should be able to swap this out with another `site` module that is focused on a specific front-end framework.  So, I'm proposing a new `JAWS-site-react` module/repo, which will contain JAWS + React boilerplate.  Here are some design considerations:

* **Keep It Simple:**  There should be little bloat in this module.  This module is for convenience, not strong opinions and obscure React patterns.  Make it the best possible starting point for a React project, that is all.
  
* **Make It Extendable**  While this module is meant for convenience, there should be some mechanism for plugging in other JAWS + React modules.  There have been lots of requests to do AWS stuff without API Gateway, since many tasks can be accomplished by using the AWS JS SDK in the browser to interact directly with AWS services.  For example, a `JAWS-site-react-s3-upload` module would be incredibly useful and fast since it would upload from the browser, directly to S3.  This module needs to also contain config files for the AWS Browser JS SDK and make them accessible to other JAWS + React modules.

* **Integrate With JAWS REST API**  JAWS will have Swagger integration finished soon.  There should be some service that allows you easily interact with your JAWS REST API.  Keeping this simple again is key.  Perhaps we can get the `JAWS-cli` to automatically scaffold this for you based on your Swagger JSON.

These are just my opinions.  Please contribute your thoughts/ideas/advice.  Also, I'm not a great React developer, so I'm looking for some better people to take the lead on this.  If you'd like to do that, please say so here :)

**Comment by SOSANA** [comment]
Add me to this list

**Comment by mcwhittemore** [comment]
@ac360 I'm definitely interested. I think site should be where all complied FE modules live so that the cli can have a simpler time of deploying the FE assets.

Have you thought much about how basic you want these modules to be? Are we talking a simple set of entry point files, are we talking a fully fledged build system?

**Comment by mbreton** [comment]
@ac360 My first idea was to build an opinionated example with some of the battle tested tools like Gulp, JQuery, Moment, Lodash, JSPM, Mocha and so on... But after read your post, I thought about it and finally I agree with you. We have to keep it most simple as possible in order to the futur adopters don't have to learn an other tool than the tool which they are interested (JAWS + React in this case).
However, about extensibility, I reminds that React is a framework for View only, and is not supposed to be responsible for the calls to the backend. So, if we want build an example simple and clean too I think, we will need a structuration framework ([Reflux](https://github.com/reflux/refluxjs) for example or an other).

**Comment by SOSANA** [comment]
opening pandora's box with flux implementation @mbreton lol this is where the real debate starts, 
@ac360 made a great comment about "Currently, I think JAWS should support multiple front-end frameworks, and each should have their own repo within the JAWS organization"

utilizing modules is the way to go under the jaws organization. I think for demo purposes simple hello with basic routes implemented, but if other maintiners want to add addional modules with implementation I think that is under consideration :), that way we have the freedom to implement any bleeding edge technologies modules with various implementations, add that my vote +1

**Comment by austinrivas** [comment]
Just finished a poc react ssr jaws demo.

Currently it is using api gateway tied to a lambda to render markup react side.

If you have react inspector installed you'll notice that the `html` `head` and `body` elements are all actually react components rendered server side and bootstrapped on the client.

The build pipeline is explained in more detail over at jaws-stack/JAWS-cli/issues/9

[Here is the live demo](https://upfxvvfdzg.execute-api.us-west-2.amazonaws.com/prod/gulp_ssr_react)
[Here is the source if anyone is looking for inspiration](https://github.com/austinrivas/gulp-lambda-boilerplate/tree/react-ssr)

It uses the same IAM and .env config as JAWS.

**Comment by ac360** [comment]
Update 8/13/15:  

We’re currently working on a document that outlines the specifications for JAWS v1.0.0.  

A proposed resolution to this issue in particular is included in that document.  

Today or tomorrow, we will share the document for everyone to comment on because your feedback is essential.

**Comment by ac360** [comment]
Outdated
